lib/postgres/postgres-pr/connection.rb

```diff
@@ -129,12 +129,101 @@ def peerport
     @conn.peerport
   end
 
+  def peerinfo
+    "#{peerhost}:#{peerport}"
+  end
+
   def current_database
     @params['database']
   end
 
-  def peerinfo
-    "#{peerhost}:#{peerport}"
+  # List of supported PostgreSQL platforms & architectures:
+  # https://postgrespro.com/docs/postgresql/16/supported-platforms
+  def map_compile_os_to_platform(compile_os)
+    return Msf::Platform::Unknown.realname if compile_os.blank?
+
+    compile_os = compile_os.downcase.encode(::Encoding::BINARY)
+
+    if compile_os.match?('linux')
+      platform = Msf::Platform::Linux
+    elsif compile_os.match?(/(darwin|mac|osx)/)
+      platform = Msf::Platform::OSX
+    elsif compile_os.match?('win')
+      platform = Msf::Platform::Windows
+    elsif compile_os.match?('free')
+      platform = Msf::Platform::FreeBSD
+    elsif compile_os.match?('net')
+      platform = Msf::Platform::NetBSD
+    elsif compile_os.match?('open')
+      platform = Msf::Platform::OpenBSD
+    elsif compile_os.match?('solaris')
+      platform = Msf::Platform::Solaris
+    elsif compile_os.match?('aix')
+      platform = Msf::Platform::AIX
+    elsif compile_os.match?('hpux')
+      platform = Msf::Platform::HPUX
+    elsif compile_os.match?('irix')
+      platform = Msf::Platform::Irix
+    else
+      # Return the query result if the value can't be mapped
+      return compile_os
+    end
+
+    platform.realname
+  end
+
+  # List of supported PostgreSQL platforms & architectures:
+  # https://postgrespro.com/docs/postgresql/16/supported-platforms
+  def map_compile_arch_to_architecture(compile_arch)
+    return '' if compile_arch.blank?
+
+    compile_arch = compile_arch.downcase.encode(::Encoding::BINARY)
+
+    if compile_arch.match?('sparc')
+      if compile_arch.include?('64')
+        arch = ARCH_SPARC64
+      else
+        arch = ARCH_SPARC
+      end
+    elsif compile_arch.include?('mips')
+      arch = ARCH_MIPS
+    elsif compile_arch.include?('ppc')
+      arch = ARCH_PPC
+    elsif compile_arch.match?('arm')
+      arch = ARCH_AARCH64
+    elsif compile_arch.match?('64')
+      arch = ARCH_X86_64
+    elsif compile_arch.match?('86') || compile_arch.match?('i686')
+      arch = ARCH_X86
+    else
+      # Return the query result if the value can't be mapped
+      arch = compile_arch
+    end
+
+    arch
+  end
+
+  # @return [Hash] Detect the platform and architecture of the MySQL server:
+  #  * :arch [String] The server architecture.
+  #  * :platform [String] The server platform.
+  def detect_platform_and_arch
+    result = {}
+
+    # I could find a clean way to grab these in a way that MySQL seemed to be able to, closest
+    # I could get was grabbing the version and then plucking out the parts I needed.
+    # Let me know if there is a better way to handle this that I may not have came across while
+    # googling.
+    query_result = query('select version()').rows.join.match(/.*on (\w+-\w+-\w+-\w+),/).captures
```

Comment:
Had to go with querying here, as the client data for the version doesn't contain the required information. It appears to be shortened.

```
[1] pry(#<Msf::Db::PostgresPR::Connection>)> self
=> #<Msf::Db::PostgresPR::Connection:0x00007fe889d5ed40
 @conn=#<Socket:fd 15>,
 @params=
  {"username"=>"postgres",
   "database"=>"template1",
   "in_hot_standby"=>"off",
   "integer_datetimes"=>"on",
   "TimeZone"=>"Etc/UTC",
   "IntervalStyle"=>"postgres",
   "is_superuser"=>"on",
   "application_name"=>"",
   "default_transaction_read_only"=>"off",
   "scram_iterations"=>"4096",
   "DateStyle"=>"ISO, MDY",
   "standard_conforming_strings"=>"on",
   "session_authorization"=>"postgres",
   "client_encoding"=>"UTF8",
   "server_version"=>"16.1 (Debian 16.1-1.pgdg120+1)", <--- HERE
   "server_encoding"=>"UTF8"},
 @transaction_status=73>
[2] pry(#<Msf::Db::PostgresPR::Connection>)> query('select version()').rows
=> [["PostgreSQL 16.1 (Debian 16.1-1.pgdg120+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit"]]
```